client/exec.go

```diff
@@ -0,0 +1,305 @@
+	if opts.Timeout != 0 {
+		timeoutStr = opts.Timeout.String()
+	}
+
+	var payload = struct {
+		Command     []string          `json:"command"`
+		Environment map[string]string `json:"environment"`
+		WorkingDir  string            `json:"working-dir"`
+		Timeout     string            `json:"timeout"`
+		UserID      *int              `json:"user-id"`
+		User        string            `json:"user"`
+		GroupID     *int              `json:"group-id"`
+		Group       string            `json:"group"`
+		Terminal    bool              `json:"terminal"`
+		Stderr      bool              `json:"stderr"`
+		Width       int               `json:"width"`
+		Height      int               `json:"height"`
+	}{
+		Command:     opts.Command,
+		Environment: opts.Environment,
+		WorkingDir:  opts.WorkingDir,
+		Timeout:     timeoutStr,
+		UserID:      opts.UserID,
+		User:        opts.User,
+		GroupID:     opts.GroupID,
+		Group:       opts.Group,
+		Terminal:    opts.Terminal,
+		Stderr:      opts.Stderr,
+		Width:       opts.Width,
+		Height:      opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return "", err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return "", err
+	}
+	var result struct {
+		WebsocketIDs map[string]string `json:"websocket-ids"`
+	}
+	err = json.Unmarshal(resultBytes, &result)
+	if err != nil {
+		return "", err
+	}
+
+	if args == nil {
+		return changeID, nil
+	}
+
+	// Process additional arguments (connecting I/O and websockets)
+	fds := result.WebsocketIDs
+
+	// Call the control handler with a connection to the control socket
+	if args.Control != nil && fds["control"] != "" {
+		conn, err := client.getChangeWebsocket(changeID, fds["control"])
+		if err != nil {
+			return "", err
+		}
+		go args.Control(conn)
+	}
+
+	if opts.Terminal {
+		// Handle terminal-based executions
+		if args.Stdin != nil && args.Stdout != nil {
+			// Connect to the websocket
+			conn, err := client.getChangeWebsocket(changeID, fds["io"])
+			if err != nil {
+				return "", err
+			}
+
+			// And attach stdin and stdout to it
+			go func() {
+				wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+				<-wsutil.WebsocketRecvStream(args.Stdout, conn)
+				conn.Close()
+				if args.DataDone != nil {
+					close(args.DataDone)
+				}
+			}()
+		} else {
+			if args.DataDone != nil {
+				close(args.DataDone)
+			}
+		}
+	} else {
+		// Handle non-terminal executions
+		dones := map[string]chan bool{}
+		conns := []*websocket.Conn{}
+
+		// Handle stdin and stdout
+		if fds["io"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["io"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stdin"] = wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+			dones["stdout"] = wsutil.WebsocketRecvStream(args.Stdout, conn)
```

Comment:
Similar to an earlier comment, I'm a bit confused about why we have different handling of stdin/stdout on the client of our API depending on whether we're dealing with a terminal or not.  The idea as I understood it was to have stdin/out be one socket, and stderr if separated out be a different socket. There's nothing about the handling of Stdin/out that should change when using a terminal or not using a terminal. Reading until here, I don't understand yet why it's tricky to not go ahead and do that entire design, but even if we implement it only partially for now, it should still be true that users of our own API should see those two cases consistently, and if it is indeed consistent the code here should reflect that by having both as a single code path. Or perhaps I am still missing something.

Reply:
You're right, there's no need for this (certainly in the client). I had copied that structure from LXD without really digging in too deep (usually a mistake :-). I've significantly simplified this now, and there's only one straight-forward code path whether or not `UseTerminal` is true. Thanks.

Reply:
Ha, I was working through it again and realized that it wasn't at all hard to implement the full combinations now -- after some cleanup it turned out to be a 3-line code change. :-) https://github.com/benhoyt/pebble/commit/26c0adccff56317b1a111622fe387bb1732caac8